MohrCoulomb plasticity has a custom return-map function

WilkAndy opened the pull request:
Fixes #6242

WilkAndy committed:
TensorMechanicsPlasticMohrCoulomb now has a custom return-map function.

Other changes:

TensorMechanicsPlasticModel and derived classes:
Apparently minor change: Kuhn-Tucker now uses the ep_plastic_tolerance to discern whether the plastic multipliers are negative or not.  I believe this is mathematically correct, ut it may have large implications on certain return-map processes.  For example, there will be some return-maps which previously were deemed invalid due to Kuhn-Tucker checks, but now they are OK.  This change could be quite noticable if the user had choosen a large ep_plastic_tolerance.  However, that's simply a physical choice a user has to make based on their specific simulation, and MOOSE can't do anything about it.

TensorMechanicsPlasticJ2 and TensorMechanicsPlasticTensileMulti
have useCustomReturnMap and useCustomCTO functions that allow functions (such as in ComputeMultiPlasticityStress) using these classes to ascertain whether they are using the custom versions or not.

Fixes #6242

WilkAndy committed:
Tests of the custom returnMap of Mohr-Coulomb.

Also a number of tests trivially re-golded so they can use the custom returnMap.

Also a number of tests modified because they had a much too large ep_plastic_tolerance.

Fixes #6242

permcody commented:
@dschwen please review

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h:
```diff
@@ -95,13 +101,12 @@ class TensorMechanicsPlasticJ2 : public TensorMechanicsPlasticModel
 
-  /// Whether to use the custom returnMap algorithm
+  /// Whether to use the custom return-map algorithm
   bool _use_custom_returnMap;
```
This should probably be const (unless it can be changed after construction). Same with `_use_custom_cto`.

WilkAndy replied:
const unless it can be changed after construction:

I believed that we couldn't (or shouldn't) change any GeneralUserObject member variable after construction.  There are times i would have liked to, but thought it was naughty!

WilkAndy replied:
Done (as well as other similar things in the .h files)

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h:
```diff
@@ -248,7 +248,14 @@ class TensorMechanicsPlasticModel : public GeneralUserObject
+   * @param dpm plastic multiplier
+   * @param dpm_tol tolerance on plastic multiplier: viz dpm>-dpm_tol means "dpm is non-negative"
+   */
+  bool KuhnTuckerSingleSurface(const Real & yf, const Real & dpm, const Real & dpm_tol) const;
```
`const Real &` can be replaced with just `Real` for improved readability. They are almost the same thing (except for constness within the function scope, but we do not care about this _anywhere_ in MOOSE).

WilkAndy replied:
Done.  Also changed "const int &" to "int" and "const bool &" to "bool".  There were about 250 of these in tensor_mechanics!

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticTensileMulti.h:
```diff
@@ -173,18 +179,18 @@ class TensorMechanicsPlasticTensileMulti : public TensorMechanicsPlasticModel
   unsigned int _max_iters;
 
   /// yield function is shifted by this amount to avoid problems with stress-derivatives at equal eigenvalues
   Real _shift;
```
This should also be const. It helps people reading the code to make clear that this member is not changed after construction, and it helps avoid bugs in the code by accidentally changing these members.

WilkAndy replied:
Done

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticTensileMulti.C:
```diff
@@ -498,10 +498,10 @@ TensorMechanicsPlasticTensileMulti::returnPlane(const std::vector<Real> & eigval
 
 bool
-TensorMechanicsPlasticTensileMulti::KuhnTuckerOK(const RankTwoTensor & returned_diagonal_stress, const std::vector<Real> & dpm, const Real & str) const
+TensorMechanicsPlasticTensileMulti::KuhnTuckerOK(const RankTwoTensor & returned_diagonal_stress, const std::vector<Real> & dpm, const Real & str, const Real & ep_plastic_tolerance) const
```
Hm... this is weird, the function signature is different here. In other classes the first argument is a `std::vector<Real>`.

WilkAndy replied:
Uncommon function arguments for KuhnTuckerOK:

This is for performance - evaluating the yield function (the std::vector thing) is expensive and not actually needed right here.

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticMohrCoulombMulti.h:
```diff
@@ -98,6 +98,9 @@ class TensorMechanicsPlasticMohrCoulombMulti : public TensorMechanicsPlasticMode
+  /// Returns _use_custom_returnMap
+  virtual bool useCustomReturnMap() const;
+
 
```
extra whitespace

WilkAndy replied:
got rid of lots of occurrences of extra wihitespaces.  obviously i like whitespace - easiest thing to read in the world :-)

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticMohrCoulombMulti.h:
```diff
@@ -110,9 +113,15 @@ class TensorMechanicsPlasticMohrCoulombMulti : public TensorMechanicsPlasticMode
+  unsigned int _max_iters;
+
   /// yield function is shifted by this amount to avoid problems with stress-derivatives at equal eigenvalues
   Real _shift;
```
const

WilkAndy replied:
Done

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticMohrCoulombMulti.h:
```diff
@@ -131,6 +140,94 @@ class TensorMechanicsPlasticMohrCoulombMulti : public TensorMechanicsPlasticMode
+   */
+  void yieldFunctionEigvals(const Real & e0, const Real & e1, const Real & e2, const Real & sinphi, const Real & cohcos, std::vector<Real> & f) const;
+
+
```
extra whitespace (and above)

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticMohrCoulombMulti.h:
```diff
@@ -158,6 +255,125 @@ class TensorMechanicsPlasticMohrCoulombMulti : public TensorMechanicsPlasticMode
+   * @param dpm The six plastic multipliers
+   * @param ep_plastic_tolerance The tolerance on the plastic strain (if dpm>-ep_plastic_tolerance then it is classified as "non-negative" in the Kuhn-Tucker conditions).
+   */
+  bool KuhnTuckerOK(const std::vector<Real> & yf, const std::vector<Real> & dpm, const Real & ep_plastic_tolerance) const;
```
this is an example for the different signature

dschwen commented on modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticMohrCoulombMulti.h:
```diff
@@ -158,6 +255,125 @@ class TensorMechanicsPlasticMohrCoulombMulti : public TensorMechanicsPlasticMode
+  bool doReturnMap(const RankTwoTensor & trial_stress, const Real & intnl_old, const RankFourTensor & E_ijkl,
+                                  Real ep_plastic_tolerance, RankTwoTensor & returned_stress, Real & returned_intnl,
+                                  std::vector<Real> & dpm, RankTwoTensor & delta_dp, std::vector<Real> & yf,
+                                  bool & trial_stress_inadmissible) const;
```
Indentation (and again `const Real &`  that should just be `Real` imo)

WilkAndy replied:
Done - and other similar indentation issues

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -17,7 +17,7 @@ InputParameters validParams<ComputeMultiPlasticityStress>()
   params.addClassDescription("Base class for multi-surface finite-strain plasticity");
   params.addRangeCheckedParam<unsigned int>("max_NR_iterations", 20, "max_NR_iterations>0", "Maximum number of Newton-Raphson iterations allowed");
-  params.addRequiredRangeCheckedParam<Real>("ep_plastic_tolerance", "ep_plastic_tolerance>0", "The Newton-Raphson process is only deemed converged if the plastic strain increment constraints have L2 norm less than this.");
+  params.addRequiredParam<Real>("ep_plastic_tolerance", "The Newton-Raphson process is only deemed converged if the plastic strain increment constraints have L2 norm less than this.");
```
Any reason you removed the range check here? Negative tolerances fro L2 norm values don't really make sense.

dschwen commented on modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C:
```diff
@@ -77,6 +77,9 @@ ComputeMultiPlasticityStress::ComputeMultiPlasticityStress(const InputParameters
 {
+  if (_epp_tol <= 0)
+    mooseError("ComputeMultiPlasticityStress: ep_plastic_tolerance must be positive");
+
```
Oh, you are checking here. But why? This is what we have range checking for.

WilkAndy replied:
Removing RangeCheckedParam in favour of an explicit check on _epp_tol in the constructor:

That tolerance is often set quite small (1E-12 is pretty standard), so the RangeChecked thing fails.  In that case i thought the explicit check was the simplest.

permcody replied:
Really? What expression are you using? Maybe we can make it better.

WilkAndy replied:
yea, when you simply write

params.addRequiredRangeCheckedParam<Real>("andy_brain_size", "andy_brain_size>0", "Should be set enormously huge");

then if someone sets "andy_brain_size=1E-12" it violates the RangeChecking, as "zero" in the above check is actually "1E-12".

permcody replied:
Hmm, @dschwen can probably tell us how to mess with tolerances in fparser. That's not something I'd considered when I built this system but clearly it's needed.

dschwen replied:
Yeah, you can call `setEpsilon(tol)` on an `FParser` object to set the tolerance for the fuzzy compare.

permcody replied:
OK, then we'll need to engineer that into the rangeCheckedParam call somehow. I'll open a ticket so we can think about it.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticModel.C:
```diff
@@ -189,15 +201,15 @@ TensorMechanicsPlasticModel::returnMap(const RankTwoTensor & trial_stress, const
   std::vector<Real> dpm(numberSurfaces(), 0);
   for (unsigned sf = 0 ; sf < numberSurfaces() ; ++sf)
-    if (!KuhnTuckerSingleSurface(yf[sf], dpm[sf]))
+    if (!KuhnTuckerSingleSurface(yf[sf], dpm[sf], 0))
```
A default of 0.0 for the dpm_tol parameter might make sense here (just an idea)

WilkAndy replied:
I've decided not to make a default value for dpm_tol in this function.  It is good to have a tolerance value entered as a non-default argument i think.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -54,18 +58,24 @@ TensorMechanicsPlasticMohrCoulombMulti::yieldFunctionV(const RankTwoTensor & str
+}
+
+void
+TensorMechanicsPlasticMohrCoulombMulti::yieldFunctionEigvals(const Real & e0, const Real & e1, const Real & e2, const Real & sinphi, const Real & cohcos, std::vector<Real> & f) const
```
More `const Real &` that should probably be just `Real`

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -190,45 +200,134 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
+  Real sinphi = std::sin(phi(intnl_old));
+  Real cosphi = std::cos(phi(intnl_old));
+  Real coh = cohesion(intnl_old);
+  Real cohcos = coh*cosphi;
```
All these "shorthand" variables should be declared const. It makes their intention more clear to the reader (and the compiler!)

WilkAndy replied:
Done, and lots more similar "const" declarations added.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-  pminusa[2] = eigvals[2] - lam1;
-  // p = a + alpha*n1 + beta*n3, and for Kuhn-Tucker we need alpha and beta non-negative
-  if (dot(pminusa, n3xn5)/dot(n[1], n3xn5) >= 0 && dot(pminusa, n1xn5)/dot(n[3], n1xn5) >= 0)
+  for (trial = 0 ; trial < 5 ; ++trial)
```
We don't do whitespaces _before_ the semicolons.

dschwen replied:
And what is the significance of `5` trials? Can this be a `const unsigned int` with a descriptive name, because then...

WilkAndy replied:
No whitespaces before semicolons?!  OK, there were 49 such lines in tensor_mechanics - now they're all fixed.

WilkAndy replied:
Replaced "5" with number_of_return_paths.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-  pminusa[2] = eigvals[2] - lam2;
-  // p = a + alpha*n3 + beta*n5, and for Kuhn-Tucker we need alpha and beta non-negative
-  if (dot(pminusa, n1xn5)/dot(n[3], n1xn5) >= 0 && dot(pminusa, n1xn3)/dot(n[5], n1xn3) >= 0)
+  if (trial == 5)
```
...this line would be more readable as well.

WilkAndy replied:
Done

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
+  for (unsigned i = 0 ; i < 6 ; ++i)
+    returned_intnl += dpm[i];
+  for (unsigned i = 0 ; i < 6 ; ++i)
+      for (unsigned j = 0 ; j < 3 ; ++j)
```
indentation

WilkAndy replied:
Done

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-TensorMechanicsPlasticMohrCoulombMulti::cross(const std::vector<Real> & a, const std::vector<Real> & b, std::vector<Real> & c) const
+
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnTip(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```
Is there a reason you are not using `RealVectorValue` here? It would seem like a more natural choice for a vector that is representing spatial components (i.e. size == 3, combined with a LIBMESH_DIM == 3 check)

WilkAndy replied:
Yep, i'll change the "n" vectors to be RealVectorValue guys.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-Real
-TensorMechanicsPlasticMohrCoulombMulti::dphi(const Real internal_param) const
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnEdge000101(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, const Real & mag_E, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```
Bender: "To quote the prophet Jer-a-matic, one zero zero zero one zero one zero one zero one zero one..." *time passes* "...zero zero one zero one one zero zero one...two. Amen."

WilkAndy replied:
Heheheh, i love that prophet.  Here, and elsewhere, the "000101" etc, tells the reader which of the 6 surfaces of the MohrCoulomb yield criterion are active.  It's very important to know this otherwise the return* functions don't make sense.  I need to highlight this in the in-code documentation.

dschwen commented on modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C:
```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
-TensorMechanicsPlasticMohrCoulombMulti::psi(const Real internal_param) const
+
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnEdge010100(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, const Real & mag_E, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```
This function looks oddly similar to `returnEdge000101`. Is there a way to generalize this maybe?

WilkAndy replied:
Generalise returnEdge000101 and returnEdge010100 to just "returnEdge".  I actually did this for returnTip110100 and returnTip010101 (and made the returnTip function) which is only a little obscure because of that generalisation.  I did also write a "returnEdge" function but it contained so many "if" statements that it was just too obscure - it's already a nightmare with those stupid Mathematica-generated expressions!  So i'll consider this, but am not that hopeful.

WilkAndy commented:
Thanks @dschwen for all your comments - i'll address them in the next iteration.

a

WilkAndy committed:
Addressed all of DSchwen comments.

Lots of minor changes involving colons, whitespace, and so on, picked up by Daniel.
I changed lots of files, but the most important changes occur in the following files:
TensorMechanicsPlasticModel
TensorMechanicsPlasticJ2
TensorMechanicsPlasticTensileMulti
TensorMechanicsPlasticTensorMohrCoulomb

Fixes #6242

permcody commented:
Wow, lots of cleanup here. Nice!

WilkAndy commented:
Thanks @dschwen and @permcody ,

a
